Extension/src/LanguageServer/cppBuildTaskProvider.ts

```diff
@@ -0,0 +1,346 @@
+    private writeEmitter  = new vscode.EventEmitter<string>();
+    private closeEmitter = new vscode.EventEmitter<number>();
+    public get onDidWrite(): vscode.Event<string> { return this.writeEmitter.event; }
+    public get onDidClose(): vscode.Event<number> { return this.closeEmitter.event; }
+
+    constructor(private command: string, private args: string[], private options: cp.ExecOptions | undefined) {
+    }
+
+
+    open(_initialDimensions: vscode.TerminalDimensions | undefined): void {
+        telemetry.logLanguageServerEvent("cppBuildTaskStarted");
+        // At this point we can start using the terminal.
+        this.writeEmitter.fire("Starting build...\r\n");
+        this.doBuild();
+    }
+
+    close(): void {
+        // The terminal has been closed. Shutdown the build.
+    }
+
+    private async doBuild(): Promise<number> {
+        return new Promise<number>((resolve, reject) => {
+            // Do build.
+            let activeCommand: string = util.resolveVariables(this.command, this.AdditionalEnvironment);
+            this.args.forEach(value => {
+                let temp: string = util.resolveVariables(value, this.AdditionalEnvironment);
+                if (temp && temp.includes(" ")) {
+                    temp = "\"" + temp + "\"";
+                }
+                activeCommand = activeCommand + " " + temp;
+            });
+            if (this.options?.cwd) {
+                this.options.cwd = util.resolveVariables(this.options.cwd, this.AdditionalEnvironment);
+            }
+            cp.exec(activeCommand, this.options, (_error, stdout, _stderr) => {
+                if (_error) {
+                    telemetry.logLanguageServerEvent("cppBuildTaskError");
+                    this.writeEmitter.fire("Build finished with error:\r\n");
+                    this.writeEmitter.fire(stdout.toString());
+                    reject();
+                } else {
+                    this.writeEmitter.fire(stdout.toString());
+                    this.writeEmitter.fire("\r\nBuild finished successfully.\r\n");
+                    resolve(0);
+                }
+            });
+        }).finally (() => {
+            this.closeEmitter.fire(0);
+        });
+    }
+
+    private get AdditionalEnvironment(): { [key: string]: string | string[] } | undefined {
+
+        const editor: vscode.TextEditor | undefined = vscode.window.activeTextEditor;
+        if (!editor) {
+            return undefined;
+        }
+        const fileDir: vscode.WorkspaceFolder | undefined = vscode.workspace.getWorkspaceFolder(editor.document.uri);
+        if (!fileDir) {
+            return undefined;
+        }
+        const file: string = editor.document.fileName;
+        return {
+            "file": file,
+            "fileDirname": fileDir.uri.fsPath,
+            "fileBasenameNoExtension": path.parse(file).name,
+            "workspaceFolder": fileDir.uri.fsPath};
```

Comment:
nit: the closing curly brace should be on a new line

Reply:
I format the whole file.